azurelinuxagent/common/cgroupstelemetry.py

```diff
@@ -13,47 +13,184 @@
-            del (CGroupsTelemetry._tracked[name])  # TODO - thread-safeness; review implementation
+        # Pop is atomic and thread-safe.
+        return CGroupsTelemetry._tracked.pop(name, None)
 
     @staticmethod
     def collect_all_tracked():
-        pass  # TODO
+        collected_metrics = {}
+        for name, cgroup_metrics in CGroupsTelemetry._cgroup_metrics.items():
+            collected_metrics[name] = CGroupsTelemetry._process_cgroup_metric(cgroup_metrics)
+            cgroup_metrics.clear()
+
+        return collected_metrics
+
+    @staticmethod
+    def poll_all_tracked():
+        CGroupsTelemetry._rlock.acquire()
+        for cgroup in CGroupsTelemetry._tracked:
+            if cgroup.name not in CGroupsTelemetry._cgroup_metrics:
+                CGroupsTelemetry._cgroup_metrics[cgroup.name] = CgroupMetrics()
+
+            CGroupsTelemetry._cgroup_metrics[cgroup.name].add_new_data(cgroup.controller,
+                                                                       cgroup.collect())
+            if not cgroup.is_active():
+                CGroupsTelemetry._tracked.remove(cgroup)
+
+        CGroupsTelemetry._rlock.release()
+
+    @staticmethod
+    def prune_all_tracked():
+        CGroupsTelemetry._rlock.acquire()
+        for cgroup in CGroupsTelemetry._tracked:
+            if not cgroup.is_active():
+                CGroupsTelemetry._tracked.remove(cgroup)
+
+        CGroupsTelemetry._rlock.release()
+
+    @staticmethod
+    def cleanup():
+        CGroupsTelemetry._tracked *= 0  # emptying the list
 
     def __init__(self):
-        pass  # TODO
+        pass
+        # TODO
+
+
+class CgroupMetrics(object):
+    def __init__(self):
+        self.current_memory_usage = Metric()
+        self.max_memory_levels = Metric()
+        self.current_cpu_usage = Metric()
+
+    def _add_memory_usage(self, metric):
+        self.current_memory_usage.append(metric[0].value)
+        self.max_memory_levels.append(metric[1].value)
+
+    def _add_cpu_usage(self, metric):
+        self.current_cpu_usage.append(metric[0].value)
 
+    def add_new_data(self, controller, metric):
+        if controller is "cpu":
+            self._add_cpu_usage(metric)
+        elif controller is "memory":
+            self._add_memory_usage(metric)
+
+    def get_metrics(self):
+        return self.current_memory_usage, self.max_memory_levels, self.current_cpu_usage
+
+    def clear(self):
+        self.current_memory_usage.clear()
+        self.max_memory_levels.clear()
+        self.current_cpu_usage.clear()
+
+
+class Metric(object):
+    def __init__(self):
+        self.data = []
+        self.first_poll_time = dt.utcnow()
+        self.last_poll_time = dt.utcnow()
+
+    def append(self, metric):
+        self.data.append(metric)
+        self.last_poll_time = dt.utcnow()
+
+    def clear(self):
+        self.data *= 0
+
+    def average(self):
+        return sum(self.data) / len(self.data)
```

Comment:
We need an explicit cast to float here if we don't want to end up with an int value and use integer division.